search: Convert to function component

sarayourfriend opened the pull request:
#### Changes proposed in this Pull Request

* Converts the `Search` component to be a function component.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Run `search:storybook:start` and test the search component.
* Open `/plugins` and ensure that search still works
* Open `/people/edit/:site/:user` where `site` is a site with more than 10 users. Scroll to the bottom and select the option to attribute works to another user and ensure search still functions.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

sarayourfriend committed:
search: Convert to function component

sarayourfriend committed:
Refactor with useImperativeHandle

noahtallen commented on packages/search/src/search.tsx:
```diff
@@ -90,349 +90,352 @@ type State = {
-	closeListener = keyListener( this.closeSearch );
-	openListener = keyListener( this.openSearch );
+let currentId = 0;
+const getUniqueId = () => currentId++;
```
Do we have access to something like https://github.com/WordPress/gutenberg/blob/master/packages/compose/src/hooks/use-instance-id/index.js?

sarayourfriend replied:
Yes! I hadn't thought about it but we can definitely use that as long the type exists for it.

noahtallen commented on packages/search/src/search.tsx:
```diff
@@ -90,349 +90,352 @@ type State = {
+let currentId = 0;
+const getUniqueId = () => currentId++;
+
+//This is fix for IE11. Does not work on Edge.
```
nit: comment spacing, but it's mostly weird that the lint check isn't asking about it 🤔

noahtallen reviewed (commented):
Nice! Normal cases test pretty well, but I noticed that the "predefined" searches weren't working. For example, on "plugins", if you click "appearance" in the search bar, it briefly shows results and then reverts back to no search.

![2021-02-12 12 10 52](https://user-images.githubusercontent.com/6265975/107817824-81955180-6d2b-11eb-848c-e5aecb94f4cf.gif)

Also didn't get a chance to really dig into the typescript/FC changes

sarayourfriend commented:
Thanks for catching that @noahtallen I hadn't even thought to test that vector.

sarayourfriend committed:
Only call onSearchProp after inital mounting

sarayourfriend committed:
Add `useInstanceId` to definition for wp/compose

sarayourfriend commented:
@noahtallen bug fixed!

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -443,7 +449,71 @@ export class Search extends React.Component< Props, State > {
+	searching: false,
+};
 
 export default withI18n( Search );
```
It's better to use the `useI18n` hook now when the component is functional. The `__` function is then not mixed into the props, but is purely internal.

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -443,7 +449,71 @@ export class Search extends React.Component< Props, State > {
+	pinned: false,
+	recordEvent: noop,
+	searching: false,
+};
```
What if the default props were merged into the `InnerSearch` param destructuring? That would make the prop types much more sensible -- you wouldn't need to do things like:
```js
recordEvent?.( 'Clicked Open Search );
const hasFocus: boolean = autoFocus ?? false;
```
Both `recordEvent` and `autoFocus` have valid values from the start and don't need a fallback or optional call.

Writing:
```js
const InnerSearch = ( {
  autoFocus = false,
  recordEvent = noop
} ) => {
  ...
}
```
makes that clear to the typechecker, too.

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -90,349 +96,349 @@ type State = {
+		keyword: defaultValue ?? '',
+		isOpen: defaultIsOpen ?? false,
+		hasFocus: autoFocus ?? false,
+	} );
```
What if we used three `useState` hooks here?
```js
const [ keyword, setKeyword ] = useState( defaultValue );
const [ isOpen, setOpen ] = useState( defaultIsOpen );
const [ hasFocus, setFocus ] = useState( autoFocus );
```
That's a more "hooky" coding style I would say -- the reducer and `setState` is a straightforward rewrite of the old-fashioned class component `setState`.

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -90,349 +96,349 @@ type State = {
+	const onChange = React.useCallback(
+		( event: ChangeEvent< HTMLInputElement > ): void => {
+			setState( {
+				keyword: event.target?.value ?? keyword,
```
Can it really happen that `event.target.value` is null-ish and the `keyword` fallback is used? The `onChange` handler is attached as `<input type="search" onChange />`. `event.target` is always there (even the type of `ChangeEvent` agrees with that) and the value is always a string.

sarayourfriend replied:
I thought that `target` could be null 🤔 But you're right that it can't according to the type so 🤷🏼‍♀️

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -90,349 +96,349 @@ type State = {
-			</div>
-		);
-	}
+	const renderStylingDiv = React.useCallback( () => {
```
None of the `render*` function needs to be wrapped/memoized with `useCallback`. They are called in a way where it doesn't matter it's a different instance on every call. We can simplify the code by writing just
```js
function renderStylingDiv() {
  ...
}
```

sarayourfriend replied:
I don't really understand why this is the case, can you elaborate more?

jsnajdr replied:
You're not passing the `render` functions to any component as props. They are merely local helper functions that are called only locally during render:
```js
function MyComponent() {
  function renderPart() {
    return <div>hello</div>;
  }

  return <div>{ renderPart() }</div>;
}
```
Memoizing this with `useCallback` doesn't provide any benefit.

sarayourfriend replied:
Oh, of course, because the function will be re-declared on each render anyway 🤦🏼‍♀️ I get it now, thanks!

> Memoizing this with useCallback doesn't provide any benefit.

It's actually worse! We're doing extra work for no good reason 😂 Good catch @jsnajdr

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -90,349 +96,349 @@ type State = {
+				{ ! hideOpenIcon && (
 					// `className` is accepted for some reason the intrisic attributes for SVG won't allow it (but it does work)
 					// eslint-disable-next-line @typescript-eslint/ban-ts-comment
 					/* @ts-ignore */
```
These comments can be all removed. They probably were true with some old types for `react` or `@wordpress/icons`, but today both `<Icon />` and `React.SVGProps` accept `className` without complaining.

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -90,349 +96,349 @@ type State = {
 
-			this.props.onSearch( this.state.keyword );
+			doSearch( keyword );
 		}
```
Another way to bypass debouncing (i.e., not wait when the keyword becomes empty) is:
```js
doSearch( keyword );
if ( keyword === '' ) {
  doSearch.flush?.();
}
```

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -90,349 +96,349 @@ type State = {
+	const doSearch: ( ( search: string ) => void ) & { cancel?: () => void } = React.useMemo(
+		() => ( delaySearch.current ? debounce( onSearchProp, delayTimeout ) : onSearchProp ),
+		[ onSearchProp, delayTimeout ]
+	);
```
Why do we need to store `delaySearchProp` into a ref?

The way the `doSearch` function and the effect that calls it is written now, it will also call `onSearchProp( keyword)` when the `onSearchProp` or `delayTimeout` prop changes. Is that desired? We should call the callback only when the search `keyword` input changes. Calling the callback again on `delayTimeout` change doesn't make sense, and on `onSearchProp` change it's questionable.

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -90,349 +96,349 @@ type State = {
+		[ onSearchProp, delayTimeout ]
+	);
+
+	useUpdateEffect( () => {
```
If I mount the component with a default value, retrieved e.g. from a URL query param:
```js
<Search defaultValue="wordpress" />
```
Then it will not call `onSearchProp( 'wordpress' )` on mount. Is that desired? It seems to me that we should trigger the search even in this case.

sarayourfriend committed:
Remove unnecessary memoization and use `useState` instead of `useReducer`

sarayourfriend committed:
Remove errant prop type annotation

sarayourfriend committed:
Stop using FunctionComponent and company

fullofcaffeine commented on packages/search/src/utils.ts:
```diff
@@ -0,0 +1,25 @@
+ *
+ * Source:
+ * https://github.com/reakit/reakit/blob/HEAD/packages/reakit-utils/src/useUpdateEffect.ts
+ *
```
Not a blocker, but I'd also add a short paragraph to exemplify a real-world case where it'd make sense to use this hook.

fullofcaffeine commented on packages/search/src/utils.ts:
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+
+/**
+ * A `React.useEffect` that will not run on the first render.
```
Nit: this is a clear description of what it does, but it doesn't answer _why_ this hook was added/is needed. A short paragraph with an example of when to use it would be nice :-)

fullofcaffeine reviewed (approved):
Code looks sound, though I'm still getting acquainted with it, so take with a grain of salt. The scenarios listed in the description all tested well. I couldn't reproduce the faulty scenario reported by @noahtallen here, so I assume you already fixed it. 

I also rebased from `trunk` and now all checks are passing 🎉

sarayourfriend committed:
Support defaultValue searching

sarayourfriend commented:
@jsnajdr I hadn't considered the use case you suggested, so I've updated the component to support it. @fullofcaffeine hopefully that also covers the comment you had about `useUpdateEffect`'s unexplained usage and I had to add a disable-reason comment for the `on-mount` effect I added to support Jarda's suggestion.

sarayourfriend committed:
Add testing-library dependencies

fullofcaffeine commented on packages/search/src/search.stories.jsx:
```diff
@@ -10,10 +10,9 @@ export default { title: 'Search', component: Search };
 			placeholder="Search..."
 			fitsContainer
-			onSearch={ ( search ) => console.log( 'Searched: ', search ) } // eslint-disable-line no-console
+			onSearch={ ( search ) => console.log( 'Searched:', search ) } // eslint-disable-line no-console
```
Nit: was the removal of the space after `Searched:` intentional?

sarayourfriend replied:
Yes, apparently `console.log` will add a space between each string argument:

```js
console.log('hello', 'world') // => "hello world"
```

sarayourfriend committed:
Add tests for search

sarayourfriend commented:
@fullofcaffeine I added some unit tests 🎉 Let me know what you think!

jsnajdr commented:
> I hadn't considered the use case you suggested, so I've updated the component to support it.

@sarayourfriend Which case do you have in mind here? Triggering search immediately after mount, with the `defaultValue` prop? In that case, we should use plain `useEffect` instead of `useUpdateEffect`, is that right?

sarayourfriend commented:
@jsnajdr Correct, and we do for that case. The issue is that we don't want to call `onSearchChange` in that case, so we use `useEffect`with an empty deps array to handle the `onMount` firing of `doSearch` to cover the case you suggested and then `useUpdateEffect` to cover the responsive cases after the initial mount.

jsnajdr commented:
> Correct, and we do for that case.

I see. There's one more thing that looks suspicious. In case of
```js
<Search defaultValue="foo" delaySearch delayTimeout={ 1000 } onSearch={ ... } />
```
the initial `doSearch( 'foo' )` will be delayed by 1000ms. But on mount, it should probably always happen immediately? There's no point debouncing the callback, because the user is not typing.

sarayourfriend committed:
Do not debounce initial search

sarayourfriend committed:
Deduplicate testing libraries

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -84,366 +99,412 @@ type Props = {
+			setIsOpen( true );
+
+			focus();
+			// no need to call `onSearchOpen` as it will be called by `onFocus` once the searcbox is focused
```
If `onSearchOpen` is now called only on focus, it means it never gets the `event` parameter. It can be removed from the type.

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -84,366 +99,412 @@ type Props = {
+		(
+			event:
+				| MouseEvent< HTMLButtonElement | HTMLInputElement >
+				| KeyboardEvent< HTMLButtonElement | HTMLInputElement >
```
This complex event type could be extracted to a separate named type. It would save quite a few repetitions.

sarayourfriend committed:
Extract complex type into alias

sarayourfriend committed:
Further remove unnecessary memoizations

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -84,366 +91,389 @@ type Props = {
-	scrollOverlay = (): void => {
-		this.overlay &&
+	const scrollOverlay = (): void => {
+		overlay &&
```
This `overlay` check is redundant, and was even in the original class component. The variable always has a `{ current }` value. It's only the `overlay.current` field that can be `null`.

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -26,13 +42,13 @@ const SEARCH_DEBOUNCE_MS = 300;
 // eslint-disable-next-line @typescript-eslint/no-empty-function
```
Do folks find the `no-empty-function` rule valuable? I see it disabled often, and it's active only on TypeScript files, not on JavaScript ones. Although it checks for things that are not TypeScript specific at all. I wouldn't mind removing it in another PR.

Now when we removed `_.noop` and declare a local `noop = () => {}` per file, there will be many more places that trigger it.

tyxla replied:
I think there's too many places where we declare empty functions to just default to empty behavior. However, ideally, we should check for existence before executing a function, and that way we should be able to avoid most of the cases when we actually need `noop`. While I don't find the rule particularly helpful, I do see it as an opportunity for devs to think about how they can improve their code to not have to rely on an empty function being declared.

jsnajdr reviewed (approved):
:+1: Thanks for the great effort you did here!

sarayourfriend committed:
Remove redundant null check
